### PR TITLE
PouchDB implementation

### PR DESCRIPTION
**Pull request by mpetrunic**
@wemeetagain I've further abstracted database logic to reduce duplicated code when adding db implementation.
I've duplicated leveldb tests to run on pouchdb as well.

There is one test failing and I cannot figure out logic even in leveldb.
When you create readable stream to search Attestations, search query is something like:
```
{
   gt: 00,
   lt: 01
}
```
And you get 64 response, while I can't figure how any key of returned values matches query above.

P.S. Unfortunately, pouchdb typings are faulty and are breaking compilation so I've removed them.

### CODE REVIEW

**Comment by wemeetagain** [comment]
:heart_eyes:  Awesome work! I'll take a deeper look here in a bit.

RE the search query logic:
We're storing data with keys prefixed with a byte signifying the "type" of data.
This is why we do pre-processing on database keys eg: `encodeKey(Bucket.attestation, attestationRoot)` for most of these database operations. Bucket is the number that's prepended.

The search query `{gt: N, lt: N+1}` is supposed to be say: query all values for keys with the prefix N.

File: src/db/interface.ts [hunk 1]
```diff
@@ -16,6 +16,7 @@ export interface DBOptions {
   name?: string;
 }
 
+
```
Comment (GregTheGreek) [inline src/db/interface.ts]: Please remove the extra white space :)

File: tsconfig.json [hunk 2]
```diff
@@ -43,7 +43,7 @@
     // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
     // "typeRoots": [],                       /* List of folders to include type definitions from. */
     // "types": [],                           /* Type declaration files to be included in compilation. */
-    // "allowSyntheticDefaultImports": true,  /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
+    "allowSyntheticDefaultImports": true,  /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
```
Comment (GregTheGreek) [inline tsconfig.json]: was this causing issues?

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
Two small things otherwise LGTM

Thanks for the contribution :)))) 👍

**Comment by GregTheGreek** [comment]
@wemeetagain 
> The search query {gt: N, lt: N+1} is supposed to be say: query all values for keys with the prefix N.

im assuming that `N+1` is just an upper bound `<` ?

**Comment by mpetrunic** [comment]
> @wemeetagain
> 
> > The search query {gt: N, lt: N+1} is supposed to be say: query all values for keys with the prefix N.
> 
> im assuming that `N+1` is just an upper bound `<` ?

Yup, I've added comments to search function to emphasize this.
Figured out why test was failing, I was using capitalized options which were no longer supported.

File: tsconfig.json (hunk 2 above)
Comment (mpetrunic) [reply tsconfig.json]: It said in docs to enable it, but as it seems, it works without.

**Review by GregTheGreek — APPROVED** [review]

File: src/db/impl/pouch.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  public stop(): Promise<void> {
+    return this.db.close();
+  }
+
+  public clean(): Promise<void> {
+    return this.db.destroy();
+  }
+
+  public async batchDelete(items: any[]): Promise<any> {
+    const deletitions = [];
```
Comment (wemeetagain) [inline src/db/impl/pouch.ts]: could you fix this mis-spelling?

**Review by wemeetagain — CHANGES_REQUESTED** [review]
Looks really nice! :raised_hands: 

Only one very minor comment about a misspelling.

A future enhancement could expose more of the configurability of Pouch, eg: exposing other adapters, but it seems like that could make things a little more complicated right now.

File: src/db/impl/abstract.ts [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    );
+    await this.batchDelete(criteria);
+  }
+
+  public async getVoluntaryExits(): Promise<VoluntaryExit[]> {
+    const data = this.search({
+      gt: encodeKey(Bucket.exit, Buffer.alloc(0)),
+      lt: encodeKey(Bucket.exit + 1, Buffer.alloc(0)),
+    });
+    return deserialize(data, VoluntaryExit).deserializedData;
```
Comment (wemeetagain) [inline src/db/impl/abstract.ts]: We'll need to loop through the data and deserialize each one, like `getAttestations` above

**Review by wemeetagain — CHANGES_REQUESTED** [review]

File: src/db/impl/abstract.ts (hunk 4 above)
Comment (mpetrunic) [reply src/db/impl/abstract.ts]: nice catch, not sure how I've missed it

File: src/db/impl/pouch.ts (hunk 3 above)
Comment (mpetrunic) [reply src/db/impl/pouch.ts]: Will address those two issues, sometimes tomorrow

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**